src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -3313,20 +3313,17 @@ trait Types
     override def isHigherKinded          = true
   }
 
-  /** Precondition: zipped params/args nonEmpty.  (Size equivalence enforced structurally.)
-   */
+  /** Precondition: `params.length == typeArgs.length > 0` (enforced structurally). */
   class AppliedTypeVar(
     _origin: Type,
     _constr: TypeConstraint,
-    zippedArgs: List[(Symbol, Type)]
+    override val params: List[Symbol],
+    override val typeArgs: List[Type]
   ) extends TypeVar(_origin, _constr) {
-    require(zippedArgs.nonEmpty, this)
-    override def params: List[Symbol] = zippedArgs map (_._1)
-    override def typeArgs: List[Type] = zippedArgs map (_._2)
+    require(params.nonEmpty && sameLength(params, typeArgs), this)
     override def safeToString: String = super.safeToString + typeArgs.map(_.safeToString).mkString("[", ", ", "]")
     override def setInst(tp: Type): this.type = super.setInst {
-      val instArgs = tp.typeArgs
-      if (sameLength(zippedArgs, instArgs) && forall3(params, typeArgs, instArgs) { (param, targ, inst) =>
+      if (corresponds3(params, typeArgs, tp.typeArgs) { (param, targ, inst) =>
```

Comment:
I would be inclined to keep using `sameLength(params, tp.typeArgs)`, as a quick-fail.

Reply:
`corresponds3` looks like misguided optimization. It's only used to check type parameters vs type arguments, but on average the number of those is not that large. Probably the subtyping checks are more expensive than a second traversal of the list. But on the other hand subtyping checks are optimized to handle common cases first. I wonder if we should try to:
1. Change `corresponds3` to check the length first
2. Remove `corresponds3` altogether (I think this pattern of `sameLenght(args1, args2)` and `forall3(args1, args2, params1)` is actually what you want - you don't need to check `sameLength(args1, params1)`.

Reply:
I opened #7646

Reply:
My estimation, and understanding, which is without benchmarks: 

Most of the time, I would check length first. Since length only processes the list, not the elements, the worst it can do is to add a low-coefficient linear term (running through the list), which is easily subsumed by the main processing.  Also, on the plus side, by checking length you are already prefetching all the list nodes (not its contents) up the cache-memory hierarchy, which you may want to.

In any case, this was a side matter, and I would have just left the code as it was on that. The main issue of this PR is, after all, to avoid creating extra lists.